internal/proxy/multi_app_conn.go

```diff
@@ -147,23 +152,38 @@ func (app *multiAppConn) killTMOnClientError() {
 		}
 	}
 
-	select {
-	case <-app.consensusConnClient.Quit():
-		if err := app.consensusConnClient.Error(); err != nil {
-			killFn(connConsensus, err, app.Logger)
-		}
-	case <-app.mempoolConnClient.Quit():
-		if err := app.mempoolConnClient.Error(); err != nil {
-			killFn(connMempool, err, app.Logger)
-		}
-	case <-app.queryConnClient.Quit():
-		if err := app.queryConnClient.Error(); err != nil {
-			killFn(connQuery, err, app.Logger)
-		}
-	case <-app.snapshotConnClient.Quit():
-		if err := app.snapshotConnClient.Error(); err != nil {
-			killFn(connSnapshot, err, app.Logger)
-		}
+	type op struct {
+		connClient stoppableClient
+		name       string
+	}
+
+	for _, client := range []op{
+		{
+			connClient: app.consensusConnClient,
+			name:       connConsensus,
+		},
+		{
+			connClient: app.mempoolConnClient,
+			name:       connMempool,
+		},
+		{
+			connClient: app.queryConnClient,
+			name:       connQuery,
+		},
+		{
+			connClient: app.snapshotConnClient,
+			name:       connSnapshot,
+		},
+	} {
+		go func(name string, client stoppableClient) {
+			client.Wait()
+			if ctx.Err() != nil {
```

Comment:
Slightly pedantic, but is there a case where the context is not canceled and the client did not error that should cause `client.Wait` to return?

Reply:
Wait returns: 
- there's a context error, which means the context was canceled, and that's normal shutdown, we should 
- the context hasn't been canceled, but the abci app returns an error, because of some kind of connection issue. Tendermint needs to shut down in a rather expeditious way (this is the current SIGTERM method, and maybe in the future we can shut down tendermint in a more graceful way.) 
- the context hasn't been canceled, but the abci app has shut down for some "normal" reason and we don't need to kill tendermint right now. We do nothing.  (e.g. `ctx.Err()` and `client.Error()` are both nil)

I believe this is covered by the current implementation.

Reply:
> the context hasn't been canceled, but the abci app has shut down for some "normal" reason and we don't need to kill tendermint right now. We do nothing. (e.g. ctx.Err() and client.Error() are both nil)

I'm not totally sure what these normal reasons are that wouldn't cause a context cancel, but I believe from the kind of strange wiring that we have at the moment that it _is_ possible.